unified-runtime/source/adapters/level_zero/v2/queue_immediate_in_order.cpp

```diff
@@ -705,31 +707,154 @@ ur_result_t ur_queue_immediate_in_order_t::enqueueWriteHostPipe(
 }
 
 ur_result_t ur_queue_immediate_in_order_t::enqueueUSMSharedAllocExp(
-    ur_usm_pool_handle_t, const size_t,
-    const ur_exp_async_usm_alloc_properties_t *, uint32_t,
-    const ur_event_handle_t *, void **, ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_usm_pool_handle_t pPool, const size_t size,
+    const ur_exp_async_usm_alloc_properties_t *pProperties,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    void **ppMem, ur_event_handle_t *phEvent) {
+  TRACK_SCOPE_LATENCY(
+      "ur_queue_immediate_in_order_t::enqueueUSMSharedAllocExp");
+
+  return enqueueUSMAllocHelper(pPool, size, pProperties, numEventsInWaitList,
+                               phEventWaitList, ppMem, phEvent,
+                               UR_USM_TYPE_SHARED);
 }
 
 ur_result_t ur_queue_immediate_in_order_t::enqueueUSMHostAllocExp(
-    ur_usm_pool_handle_t, const size_t,
-    const ur_exp_async_usm_alloc_properties_t *, uint32_t,
-    const ur_event_handle_t *, void **, ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_usm_pool_handle_t pPool, const size_t size,
+    const ur_exp_async_usm_alloc_properties_t *pProperties,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    void **ppMem, ur_event_handle_t *phEvent) {
+  TRACK_SCOPE_LATENCY("ur_queue_immediate_in_order_t::enqueueUSMHostAllocExp");
+
+  return enqueueUSMAllocHelper(pPool, size, pProperties, numEventsInWaitList,
+                               phEventWaitList, ppMem, phEvent,
+                               UR_USM_TYPE_HOST);
 }
 
 ur_result_t ur_queue_immediate_in_order_t::enqueueUSMFreeExp(
-    ur_usm_pool_handle_t, void *, uint32_t, const ur_event_handle_t *,
-    ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+    ur_usm_pool_handle_t, void *pMem, uint32_t numEventsInWaitList,
+    const ur_event_handle_t *phEventWaitList, ur_event_handle_t *phEvent) {
+  TRACK_SCOPE_LATENCY("ur_queue_immediate_in_order_t::enqueueUSMFreeExp");
+  auto commandListLocked = commandListManager.lock();
+  ur_event_handle_t internalEvent = nullptr;
+  if (phEvent == nullptr) {
+    phEvent = &internalEvent;
+  }
+
+  auto zeSignalEvent = getSignalEvent(commandListLocked, phEvent,
+                                      UR_COMMAND_ENQUEUE_USM_FREE_EXP);
+  auto [pWaitEvents, numWaitEvents] =
+      getWaitListView(commandListLocked, phEventWaitList, numEventsInWaitList);
+
+  umf_memory_pool_handle_t hPool = umfPoolByPtr(pMem);
+  if (!hPool) {
+    return UR_RESULT_ERROR_INVALID_MEM_OBJECT;
+  }
+
+  UsmPool *usmPool = nullptr;
+  auto ret = umfPoolGetTag(hPool, (void **)&usmPool);
+  if (ret != UMF_RESULT_SUCCESS || !usmPool) {
+    // This should never happen
+    UR_LOG(ERR, "enqueueUSMFreeExp: invalid pool tag");
+    return UR_RESULT_ERROR_UNKNOWN;
+  }
+
+  size_t size = umfPoolMallocUsableSize(hPool, pMem);
+  (*phEvent)->RefCount.increment();
```

Comment:
I think this might now never get freed when output event was != null, because the event starts with refcount 1, and then we bump it up to 2. Afterwards, we places it in the async pool, and decrement it once after the associated object is allocated. However, best I can tell, nothing will decrement it a second time.

Reply:
If output event != null, then one copy goes to the caller, and one copy goes into async pool, so refcount should be 2, shouldn't it? As you said, refcount is decremented after taking it out of the async pool, and another decrementation should be done by the caller.

Reply:
oh, sorry, I meant when output event is null.

Reply:
Few lines later there is:
```
  if (internalEvent != nullptr) {
    internalEvent->release();
  }
```  

But still, a good point, as it is kind of redundant. I will just change it to increment refcount conditionally.